gapic-common/lib/gapic/grpc/service_stub/channel_pool.rb

```diff
@@ -0,0 +1,113 @@
+        if credentials.is_a? ::GRPC::Core::Channel
+          raise ArgumentError, "Can't create a channel pool with GRPC::Core::Channel as credentials"
+        end
+
+        @grpc_stub_class = grpc_stub_class
+        @endpoint = endpoint
+        @credentials = credentials
+        @channel_args = channel_args
+        @interceptors = interceptors
+        @config = config || Configuration.new
+
+        yield @config if block_given?
+
+        @channels = (1..@config.channel_count).map { create_channel }
+      end
+
+      ##
+      # Creates a new channel.
+      def create_channel
+        Channel.new @grpc_stub_class, endpoint: @endpoint, credentials: @credentials, channel_args: @channel_args,
+                    interceptors: @interceptors, on_channel_create: @config.on_channel_create
+      end
+
+      ##
+      # Invoke the specified RPC call.
+      #
+      # @param method_name [Symbol] The RPC method name.
+      # @param request [Object] The request object.
+      # @param options [Gapic::CallOptions, Hash] The options for making the RPC call. A Hash can be provided to
+      #   customize the options object, using keys that match the arguments for {Gapic::CallOptions.new}. This object
+      #   should only be used once.
+      #
+      # @yield [response, operation] Access the response along with the RPC operation.
+      # @yieldparam response [Object] The response object.
+      # @yieldparam operation [::GRPC::ActiveCall::Operation] The RPC operation for the response.
+      #
+      # @return [Object] The response object.
+      #
+      def call_rpc method_name, request, options: nil, &block
+        case @config.channel_selection.downcase.to_sym
+        when :least_loaded
+          channel = least_loaded_channel
+          channel.call_rpc method_name, request, options: options, &block
+        end
+      end
+
+
+      private
+
+      ##
+      # Return the least loaded channel in the pool
+      #
+      # @return [::Grpc::ServiceStub::Channel]
+      #
+      def least_loaded_channel
+        @channels.min_by(&:concurrent_streams)
+      end
+
+      # @!attribute [rw] channel_count
+      #  The number of channels in the channel pool.
+      #  return [Integer]
+      # @!attribute [rw] on_channel_create
+      #  Proc to run at the end of each channel initialization.
+      #  Proc is provided ::Gapic::ServiceStub::Channel object as input.
+      #  return [Proc]
+      # @!attribute [rw] channel_selection
+      #  The algorithm for selecting a channel for an RPC.
+      #  return [Symbol]
+      #
+      class Configuration
+        extend ::Gapic::Config
+
+        config_attr :channel_count, 1, ::Integer
+        config_attr :on_channel_create, nil, ::Proc
+        config_attr :channel_selection, :least_loaded do |value|
+          allowed = [:least_loaded]
+          allowed.any? { |klass| klass === value }
```

Comment:
The name `klass` is confusing. The current contents of `allowed` are symbols, not classes.

But otherwise, this block is not necessary. Pass the allowed values at the end of `config_attr` like this:

    config_attr :channel_selection, :least_loaded, :least_loaded

The second argument is the initial (default) value, and all subsequent arguments are valid values or types (tested using `===`).

Reply:
Thanks! 
Removed the block and sending the valid_values.